commit 261346d80a738c0f9148614aeedb4cc18b92027c (refs/changes/28/2928/1)
Author: Pedro Palazon Candel <pedro@joyent.com>
Date:   2017-11-09T18:12:36+01:00 (1 year, 11 months ago)
    
    TOOLS-1925 sdcadm binder instances created after HN setup don't work with sapi-url being a domain

diff --git a/lib/post-setup/ha-binder.js b/lib/post-setup/ha-binder.js
index 8443b17..b5e5a6f 100644
--- a/lib/post-setup/ha-binder.js
+++ b/lib/post-setup/ha-binder.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2016, Joyent, Inc.
+ * Copyright (c) 2017, Joyent, Inc.
  */
 
 /*
@@ -126,6 +126,34 @@ function do_ha_binder(subcmd, opts, args, cb) {
         // XXX vestigial?
         shared.getUserScript, // sets `arg.userScript`.
 
+        function getSapiVm(ctx, next) {
+            self.progress('Getting SDC\'s SAPI vm from VMAPI');
+            self.sdcadm.vmapi.listVms({
+                'tag.smartdc_role': 'sapi',
+                state: 'running'
+            }, function (vmsErr, vms_) {
+                if (vmsErr) {
+                    next(vmsErr);
+                    return;
+                }
+                if (!vms_.length) {
+                    next(new errors.InternalError(
+                        'Cannot find SAPI vm from VMAPI'));
+                    return;
+                }
+
+                var adminNic = vms_[0].nics.filter(function checkNic(nic) {
+                    return (nic.nic_tag === 'admin');
+                });
+                if (!adminNic.length) {
+                    next(new errors.InternalError(
+                        'Cannot find SAPI vm\'s admin Nic'));
+                    return;
+                }
+                ctx.sapiAdminIp = adminNic[0].ip;
+                next();
+            });
+        },
         function getBinderInstances(_, next) {
             if (!svc) {
                 return next();
@@ -560,7 +588,7 @@ function do_ha_binder(subcmd, opts, args, cb) {
             fs.rename(path.join(wrkDir, vms[0].server_uuid), fname, next);
         },
 
-        function createBinderInstances(_, next) {
+        function createBinderInstances(ctx, next) {
             if (instances && instances.length === (opts.members + 1)) {
                 return next();
             }
@@ -574,7 +602,8 @@ function do_ha_binder(subcmd, opts, args, cb) {
                             alias: change.inst.alias
                         },
                         metadata: {
-                            ZK_ID: String(change.inst.zk_id)
+                            ZK_ID: String(change.inst.zk_id),
+                            'sapi-url': 'http://' + ctx.sapiAdminIp
                         }
                     };
 
